src/Shared/FileMatcher.cs

```diff
@@ -1094,215 +1117,245 @@ private static string RegularExpressionFromFileSpec
             out bool isLegalFileSpec
         )
         {
-            isLegalFileSpec = true;
-
-            /*
-             * The code below uses tags in the form <:tag:> to encode special information
-             * while building the regular expression.
-             * 
-             * This format was chosen because it's not a legal form for filespecs. If the
-             * filespec comes in with either "<:" or ":>", return isLegalFileSpec=false to
-             * prevent intrusion into the special processing.
-             */
-            if ((fixedDirectoryPart.IndexOf("<:", StringComparison.Ordinal) != -1) ||
-                (fixedDirectoryPart.IndexOf(":>", StringComparison.Ordinal) != -1) ||
-                (wildcardDirectoryPart.IndexOf("<:", StringComparison.Ordinal) != -1) ||
-                (wildcardDirectoryPart.IndexOf(":>", StringComparison.Ordinal) != -1) ||
-                (filenamePart.IndexOf("<:", StringComparison.Ordinal) != -1) ||
-                (filenamePart.IndexOf(":>", StringComparison.Ordinal) != -1))
+            isLegalFileSpec = IsLegalFileSpec(wildcardDirectoryPart, filenamePart);
+            if (!isLegalFileSpec)
             {
-                isLegalFileSpec = false;
-                return String.Empty;
+                return string.Empty;
             }
 
-            /*
-             * Its not legal for there to be a ".." after a wildcard.
-             */
-            if (wildcardDirectoryPart.Contains(dotdot))
+            var matchFileExpression = StringBuilderCache.Acquire(FileSpecRegexInitialBufferSize);
```

Comment:
You could use our own ReuseableStringBuilder? That way when we replace it we replace it everywhere consistently :)